tree.go

```diff
@@ -1,170 +1,134 @@
 package ipldgit
 
 import (
+	"bufio"
 	"bytes"
-	"encoding/json"
-	"errors"
 	"fmt"
 	"io"
-	"sync"
+	"strconv"
 
-	cid "github.com/ipfs/go-cid"
-	node "github.com/ipfs/go-ipld-format"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
 )
 
-type Tree struct {
-	entries     map[string]*TreeEntry
-	size        int
-	order       []string
-	cid         cid.Cid
-	rawData     []byte
-	rawDataOnce sync.Once
-}
-
-type TreeEntry struct {
-	name string
-	Mode string  `json:"mode"`
-	Hash cid.Cid `json:"hash"`
-}
-
-func (t *Tree) Cid() cid.Cid {
-	return t.cid
-}
-
-func (t *Tree) String() string {
-	return "[git tree object]"
-}
-
-func (t *Tree) GitSha() []byte {
-	return cidToSha(t.cid)
-}
-
-func (t *Tree) Copy() node.Node {
-	out := &Tree{
-		entries: make(map[string]*TreeEntry),
-		cid:     t.cid,
-		size:    t.size,
-		order:   t.order, // TODO: make a deep copy of this
+// DecodeTree fills a NodeAssembler (from `Type.Tree__Repr.NewBuilder()`) from a stream of bytes
+func DecodeTree(na ipld.NodeAssembler, rd *bufio.Reader) error {
+	lstr, err := rd.ReadString(0)
+	if err != nil {
+		return err
 	}
+	lstr = lstr[:len(lstr)-1]
 
-	for k, v := range t.entries {
-		nv := *v
-		out.entries[k] = &nv
+	_, err = strconv.Atoi(lstr)
+	if err != nil {
+		return err
 	}
-	return out
-}
-
-func (t *Tree) MarshalJSON() ([]byte, error) {
-	return json.Marshal(t.entries)
-}
 
-func (t *Tree) Tree(p string, depth int) []string {
-	if p != "" {
-		_, ok := t.entries[p]
-		if !ok {
-			return nil
+	t := Type.Tree__Repr.NewBuilder()
```

Comment:
I'm not sure I follow why we have a Tree__Repr builder here, and its result is built and assigned into a Tree builder, which is then again built. It seems like the other decode funcs follow the same pattern. Wouldn't a single Build per Decode be enough?

Reply:
i think this came out of tests:
the default is to use a basic.any as the assembler, and if that was directly used to make a list / map you end up with a top level basic node rather than the schema'd type that we have in this package.
When that's true, the decode path can't fast-path, and has to re assign that basic node into the schema type.

Reply:
I think I agree with mvdan here -- it's a little odd to have a concrete reference like this in a method that is already accepting a NodeAssembler as a parameter.  When there is a NodeAssembler as a parameter, I would usually expect it to just be used directly.  (And maybe documentation can state something about "it is recommended to use a Tree__ReprAssembler here for best results".)

This doesn't seem dangerously incorrect as written, but does seem a little baroque.